### PR TITLE
Fix/firmware file override

### PR DESCRIPTION
**Pull request by plbossart**

### CODE REVIEW

File: sound/soc/sof/sof-pci-dev.c [hunk 1]
```diff
@@ -210,9 +214,21 @@ int sof_pci_probe(struct pci_dev *pci, const struct pci_device_id *pci_id)
...
+	 */
+	if (tplg_filename) {
+		sof_pdata->tplg_filename = tplg_filename;
+
+		dev_dbg(dev,
+			"Module parameter used, changed tplg filename to %s\n",
+			sof_pdata->tplg_filename);
+	} else {
+		dmi_check_system(sof_tplg_table);
+		if (sof_override_tplg_name)
```
Comment (ranj063) [inline sound/soc/sof/sof-pci-dev.c]: @plbossart now that we have 2 methods to override the default filename, would it make sense to rename this as sof_dmi_override_tplg_name() or something similar?

File: sound/soc/sof/sof-audio.c [hunk 2]
```diff
@@ -702,7 +702,8 @@ int sof_machine_check(struct snd_sof_dev *sdev)
 		return -ENOMEM;
 
 	mach->drv_name = "sof-nocodec";
-	sof_pdata->tplg_filename = desc->nocodec_tplg_filename;
+	if (!sof_pdata->tplg_filename)
```
Comment (ranj063) [inline sound/soc/sof/sof-audio.c:705]: typo in commit title: preserve

**Review by cujomalainey — DISMISSED** [review]
approval assuming @ranj063 comments are addressed

@plbossart thanks for this change, this is much appreciated, but this still leaves us with an issue regarding selecting non-standard topologies for boards that have different hw layouts but match regular machines in the DMI check. The OEM name hijack is still kinda gross in my opinion.

**Comment by plbossart** [comment]
> @plbossart thanks for this change, this is much appreciated, but this still leaves us with an issue regarding selecting non-standard topologies for boards that have different hw layouts but match regular machines in the DMI check. The OEM name hijack is still kinda gross in my opinion.

@cujomalainey the problem is in your court: we added a DMI string at the request of the Chrome team, if it's not used in the right way by the Chrome OEMs there's not much I can do. Feel free to suggest something else, that's orthogonal to this PR.

**Review by ranj063 — DISMISSED** [review]

File: sound/soc/sof/sof-pci-dev.c [hunk 3]
```diff
@@ -23,6 +23,10 @@ static char *fw_path;
 module_param(fw_path, charp, 0444);
 MODULE_PARM_DESC(fw_path, "alternate path for SOF firmware.");
 
+static char *fw_filename;
+module_param(fw_filename, charp, 0444);
+MODULE_PARM_DESC(fw_filename, "alternate path for SOF firmware name.");
```
Comment (paulstelian97) [inline sound/soc/sof/sof-pci-dev.c]: I'd rephrase this as "alternate filename for SOF firmware.".

File: sound/soc/sof/sof-pci-dev.c [hunk 4]
```diff
@@ -31,17 +31,21 @@ static char *tplg_path;
 module_param(tplg_path, charp, 0444);
 MODULE_PARM_DESC(tplg_path, "alternate path for SOF topology.");
 
+static char *tplg_filename;
+module_param(tplg_filename, charp, 0444);
+MODULE_PARM_DESC(tplg_filename, "alternate path for SOF firmware name.");
```
Comment (paulstelian97) [inline sound/soc/sof/sof-pci-dev.c]: Topology filename?

File: sound/soc/sof/sof-pci-dev.c [hunk 5]
```diff
@@ -160,7 +164,16 @@ int sof_pci_probe(struct pci_dev *pci, const struct pci_device_id *pci_id)
...
 
 	sof_pdata->ipc_type = desc->ipc_default;
-	sof_pdata->fw_filename = desc->default_fw_filename[sof_pdata->ipc_type];
+
+	if (fw_filename) {
+		sof_pdata->fw_filename = fw_filename;
+
+		dev_dbg(dev,
+			"Module parameter used, changed fw filename to %s\n",
+			sof_pdata->fw_filename);
```
Comment (paulstelian97) [inline sound/soc/sof/sof-pci-dev.c:176]: The first two lines should fit easily in 100 characters, no need to split.

**Review by paulstelian97 — CHANGES_REQUESTED** [review]

File: sound/soc/sof/sof-pci-dev.c (hunk 3 above)
Comment (plbossart) [reply sound/soc/sof/sof-pci-dev.c]: yes, partial copy/paste, will fix.

File: sound/soc/sof/sof-pci-dev.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/sof-pci-dev.c]: same, partial copy paste, will fix.

File: sound/soc/sof/sof-pci-dev.c (hunk 3 above)
Comment (plbossart) [reply sound/soc/sof/sof-pci-dev.c]: done

File: sound/soc/sof/sof-pci-dev.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/sof-pci-dev.c]: done

**Review by paulstelian97 — DISMISSED** [review]
The comment with the unnecessary line break still applies, but I'm fine with letting that one slide (esp. if others disagree)

**Comment by plbossart** [comment]
> The comment with the unnecessary line break still applies, but I'm fine with letting that one slide (esp. if others disagree)

I didn't change it since the rest of the code uses the same newline.

File: sound/soc/sof/sof-pci-dev.c [hunk 6]
```diff
@@ -23,6 +23,10 @@ static char *fw_path;
 module_param(fw_path, charp, 0444);
```
Comment (kv2019i) [inline sound/soc/sof/sof-pci-dev.c:23]: minor nitpick: I didn't understand "The firmware path can be used" actually refers to the "fw_path" module variable, so the rest of commit message didn't make sense at first. Using a firmware path to load firmware files seems kind of obvious until you realize it's a mechanism to use a different path. But really optional, people will understand after they look at the code.

File: sound/soc/sof/sof-pci-dev.c [hunk 7]
```diff
@@ -31,17 +31,21 @@ static char *tplg_path;
 module_param(tplg_path, charp, 0444);
```
Comment (kv2019i) [inline sound/soc/sof/sof-pci-dev.c:31]: Minor nitpicks continue. A "topology path can be used to load" ... "firmware files"?

File: sound/soc/sof/sof-pci-dev.c [hunk 8]
```diff
@@ -210,9 +214,21 @@ int sof_pci_probe(struct pci_dev *pci, const struct pci_device_id *pci_id)
...
+	/*
+	 * the topology filename will be provided in the machine descriptor, unless
+	 * it is overridden by a module parameter or DMI quirk.
+	 */
+	if (tplg_filename) {
+		sof_pdata->tplg_filename = tplg_filename;
+
+		dev_dbg(dev,
+			"Module parameter used, changed tplg filename to %s\n",
+			sof_pdata->tplg_filename);
```
Comment (kv2019i) [inline sound/soc/sof/sof-pci-dev.c:224]: Could be fit in two lines.

**Review by kv2019i — DISMISSED** [review]
A few nitpicks that you can fix or not fix based on your judgement. A very useful addition in any case.

File: sound/soc/sof/sof-pci-dev.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/sof/sof-pci-dev.c:23]: reworded.

File: sound/soc/sof/sof-pci-dev.c (hunk 7 above)
Comment (plbossart) [reply sound/soc/sof/sof-pci-dev.c:31]: reworded.

File: sound/soc/sof/sof-pci-dev.c (hunk 8 above)
Comment (plbossart) [reply sound/soc/sof/sof-pci-dev.c:224]: done

**Review by ranj063 — APPROVED** [review]

**Review by dbaluta — APPROVED** [review]

**Comment by dbaluta** [comment]
@paulstelian97 please send the equivalent PR for sof-of-dev.

**Review by paulstelian97 — APPROVED** [review]
Looks good, will submit my equivalent independently tomorrow.

**Review by ujfalusi — APPROVED** [review]
